pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,240 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	registerapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// RegisterCallbackFn is the type of the callback function that handlers will provide
+type RegisterCallbackFn func(pluginName string, versions []string, socketPath string) error
+
+// Watcher is the plugin watcher
+type Watcher struct {
+	path     string
+	handlers map[string]RegisterCallbackFn
+	stopCh   chan interface{}
+	fs       utilfs.Filesystem
+	watcher  *fsnotify.Watcher
+	wg       sync.WaitGroup
+	mutex    sync.Mutex
+}
+
+// NewWatcher provides a new watcher
+func NewWatcher(sockDir string) Watcher {
+	return Watcher{
+		path:     sockDir,
+		handlers: make(map[string]RegisterCallbackFn),
+		fs:       &utilfs.DefaultFs{},
+	}
+}
+
+// AddHandler registers a callback to be invoked for a particular type of plugin
+func (w *Watcher) AddHandler(handlerType string, handlerCbkFn RegisterCallbackFn) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	w.handlers[handlerType] = handlerCbkFn
+}
+
+// Creates the plugin directory, if it doesn't already exist.
+func (w *Watcher) createPluginDir() error {
+	glog.V(4).Infof("Ensuring Plugin directory at %s ", w.path)
+	if err := w.fs.MkdirAll(w.path, 0755); err != nil {
+		return fmt.Errorf("error (re-)creating driver directory: %s", err)
+	}
+	return nil
+}
+
+// Walks through the plugin directory to discover any existing plugin sockets.
+func (w *Watcher) traversePluginDir() error {
+	files, err := w.fs.ReadDir(w.path)
+	if err != nil {
+		return fmt.Errorf("error reading the plugin directory: %v", err)
+	}
+	for _, f := range files {
+		// Currently only supports flat fs namespace under the plugin directory.
+		// TODO: adds support for hierarchical fs namespace.
+		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
+			go w.registerPlugin(path.Join(w.path, f.Name()))
+		}
+	}
+	return nil
+}
+
+func (w *Watcher) init() error {
+	if err := w.createPluginDir(); err != nil {
+		return err
+	}
+	return w.traversePluginDir()
+}
+
+func (w *Watcher) registerPlugin(socketPath string) error {
+	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
+	glog.V(4).Infof("registerPlugin called for socketPath: %s", socketPath)
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		return fmt.Errorf("dial failed at socket %s, err: %v", socketPath, err)
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
```

Comment:
It would be great to be able to not call the driver directly, because it kind of makes it a kubernetes specific and one of goals of CSI for the driver to be completely CO agnostic, but instead to use an interface, which can be generic f definition but underneath with CSI or another plugin specific implementation.  Example, to be able to reply to GetInfo call, all CSI drivers will have to be aware of this type of request which is kind of kubernetes specific and it goes a bit against CSI approach. If registerPlugin is implemented as interface, and then each plugin uses plugin specific way to get required information, then this code is kept generic but we do not loose flexibility.
@saad-ali @vladimirvivien Would you agree  with proposed modification?

Reply:
I thought CSI sidecar container will handle the interaction with kubelet, which would also include supporting this registerapi. Can some CSI folks double verify this?

Reply:
@jiayingz If side car handles `GetInfo` request then it would require two sockets, one opened by CSI driver for CSI related communication and one opened by sidecar to reply to GetInfo request,  how do you plan to deal now with two opened sockets? Where would sidecar create `GetInfo` socket?

Reply:
Defining registration interface, imho would be the cleanest solution.

Reply:
After discussion within CSI WG, please disregards these comments for now. In the interest of time, this PR should be merged and revisited later if required.